engines/lab/special.cpp

```diff
@@ -0,0 +1,482 @@
+	else if (bridge && clean)
+		filename += '6';
+	else if (clean)
+		filename += '7';
+	else if (bridge && dirty && news)
+		filename += '4';
+	else if (dirty && news)
+		filename += '5';
+	else if (bridge && dirty)
+		filename += '2';
+	else if (dirty)
+		filename += '3';
+	else if (bridge)
+		filename += '1';
+
+	_journalText = _resource->getText(filename.c_str());
+	_journalTextTitle = _resource->getText("Lab:Rooms/jt");
+
+	Common::File *journalFile = _resource->openDataFile("P:JImage");
+	Utils *utils = _utils;
+	_journalButtonList.push_back(_event->createButton( 80, utils->vgaScaleY(162) + utils->svgaCord(1), 0, VKEY_LTARROW, new Image(journalFile, this), new Image(journalFile, this)));	// back
+	_journalButtonList.push_back(_event->createButton(194, utils->vgaScaleY(162) + utils->svgaCord(1), 2,            0, new Image(journalFile, this), new Image(journalFile, this)));	// cancel
+	_journalButtonList.push_back(_event->createButton(144, utils->vgaScaleY(164) - utils->svgaCord(1), 1, VKEY_RTARROW, new Image(journalFile, this), new Image(journalFile, this)));	// forward
+	delete journalFile;
+
+	_anim->_noPalChange = true;
+	_journalBackImage->_imageData = new byte[_graphics->_screenWidth * _graphics->_screenHeight];
+	_graphics->readPict("P:Journal.pic", true, false, _journalBackImage->_imageData);
+	_anim->_noPalChange = false;
+
+	// Keep a copy of the blank journal
+	_blankJournal = new byte[_graphics->_screenWidth * _graphics->_screenHeight];
+	memcpy(_blankJournal, _journalBackImage->_imageData, _graphics->_screenWidth * _graphics->_screenHeight);
+
+	_screenImage->_imageData = _graphics->getCurrentDrawingBuffer();
+}
+
+void LabEngine::drawJournalText() {
+	uint16 drawingToPage = 1;
+	int charsDrawn = 0;
+	const char *curText = _journalText.c_str();
+
+	while (drawingToPage < _journalPage) {
+		_music->updateMusic();
+		curText = (char *)(_journalText.c_str() + charsDrawn);
+		charsDrawn += _graphics->flowText(_journalFont, -2, 2, 0, false, false, false, false, _utils->vgaRectScale(52, 32, 152, 148), curText);
+
+		_lastPage = (*curText == 0);
+
+		if (_lastPage)
+			_journalPage = (drawingToPage / 2) * 2;
+		else
+			drawingToPage++;
+	}
+
+	if (_journalPage <= 1) {
+		curText = _journalTextTitle.c_str();
+		_graphics->flowTextToMem(_journalBackImage, _journalFont, -2, 2, 0, false, true, true, true, _utils->vgaRectScale(52, 32, 152, 148), curText);
+	} else {
+		curText = (char *)(_journalText.c_str() + charsDrawn);
+		charsDrawn += _graphics->flowTextToMem(_journalBackImage, _journalFont, -2, 2, 0, false, false, false, true, _utils->vgaRectScale(52, 32, 152, 148), curText);
+	}
+
+	_music->updateMusic();
+	curText = (char *)(_journalText.c_str() + charsDrawn);
+	_lastPage = (*curText == 0);
+	_graphics->flowTextToMem(_journalBackImage, _journalFont, -2, 2, 0, false, false, false, true, _utils->vgaRectScale(171, 32, 271, 148), curText);
+
+	curText = (char *)(_journalText.c_str() + charsDrawn);
+	_lastPage = (*curText == 0);
```

Comment:
These lines look suspicious. charsDrawn wasn't updated since the last time curText and _lastPage were set, so this doesn't do anything?

(Also, why all the casts to char* ?)